src/audio/pcm_converter/pcm_converter_generic.c

```diff
@@ -0,0 +1,160 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2019 Intel Corporation. All rights reserved.
+//
+// Author: Tomasz Lauda <tomasz.lauda@linux.intel.com>
+
+/**
+ * \file audio/pcm_converter/pcm_converter_generic.c
+ * \brief PCM converter generic processing implementation
+ * \authors Tomasz Lauda <tomasz.lauda@linux.intel.com>
+ */
+
+#include <sof/audio/buffer.h>
+#include <sof/audio/pcm_converter.h>
+#include <sof/common.h>
+#include <ipc/stream.h>
+#include <stddef.h>
+#include <stdint.h>
+
+static void pcm_copy_s16(struct comp_buffer *source, struct comp_buffer *sink,
+			 uint32_t samples)
+{
+	uint32_t buff_frag = 0;
+	int16_t *src;
+	int16_t *dst;
+	uint32_t i;
+
+	for (i = 0; i < samples; i++) {
+		src = buffer_read_frag_s16(source, buff_frag);
+		dst = buffer_write_frag_s16(sink, buff_frag);
+		*dst = *src;
+		buff_frag++;
+	}
+}
+
+static void pcm_copy_s32(struct comp_buffer *source, struct comp_buffer *sink,
+			 uint32_t samples)
+{
+	uint32_t buff_frag = 0;
+	int32_t *src;
+	int32_t *dst;
+	uint32_t i;
+
+	for (i = 0; i < samples; i++) {
+		src = buffer_read_frag_s32(source, buff_frag);
+		dst = buffer_write_frag_s32(sink, buff_frag);
+		*dst = *src;
+		buff_frag++;
+	}
+}
+
+static void pcm_convert_s16_to_s24(struct comp_buffer *source,
+				   struct comp_buffer *sink, uint32_t samples)
+{
+	uint32_t buff_frag = 0;
+	int16_t *src;
+	int32_t *dst;
+	uint32_t i;
+
+	for (i = 0; i < samples; i++) {
+		src = buffer_read_frag_s16(source, buff_frag);
+		dst = buffer_write_frag_s32(sink, buff_frag);
+		*dst = *src << 8;
+		buff_frag++;
+	}
+}
+
+static void pcm_convert_s16_to_s32(struct comp_buffer *source,
+				   struct comp_buffer *sink, uint32_t samples)
+{
+	uint32_t buff_frag = 0;
+	int16_t *src;
+	int32_t *dst;
+	uint32_t i;
+
+	for (i = 0; i < samples; i++) {
+		src = buffer_read_frag_s16(source, buff_frag);
+		dst = buffer_write_frag_s32(sink, buff_frag);
+		*dst = *src << 16;
+		buff_frag++;
+	}
+}
+
+static void pcm_convert_s24_to_s16(struct comp_buffer *source,
+				   struct comp_buffer *sink, uint32_t samples)
+{
+	uint32_t buff_frag = 0;
+	int32_t *src;
+	int16_t *dst;
+	uint32_t i;
+
+	for (i = 0; i < samples; i++) {
+		src = buffer_read_frag_s32(source, buff_frag);
+		dst = buffer_write_frag_s16(sink, buff_frag);
+		*dst = *src >> 8;
```

Comment:
Rounding is missing, also remember to saturate when rounding. Format.h contains sat_int16() function.  Macro Q_SHIFT_RND() could be used for the rounding,